### PR TITLE
Move predict methods to a trait and fix tfnet memory leak

### PR DESCRIPTION
**Pull request by yangw1234**
1. move predict methods to a trait so that other mode (like TFNet)l can use these methods
2. create a new predictor for the use of analytics-zoo
3. fix memory leak when using TFNet (create a Iterator that will call a shutdown hook when hasNext return false)

### CODE REVIEW

**Comment by yangw1234** [comment]
jekins pass

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+          i += 1
+        }
+      })
+      tables
+    }
+    out.asInstanceOf[Array[Activity]]
+  }
+
+
+  def predictImage[T: ClassTag](imageFrame: DistributedImageFrame,
```
[the remark this answers is not in the document]
Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala]: They needs to extends Predictable

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextPredictor.scala [hunk 2]
```diff
@@ -22,12 +22,11 @@ import com.intel.analytics.bigdl.models.utils.ModelBroadcast
 import com.intel.analytics.bigdl.tensor.Tensor
 import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
 import com.intel.analytics.zoo.pipeline.api.keras.layers.utils.EngineRef
-import com.intel.analytics.zoo.pipeline.api.keras.models.KerasNet
 
 import scala.reflect.ClassTag
 
 class TextPredictor[T: ClassTag](
-    model: KerasNet[T],
+    model: Module[T],
```
[the remark this answers is not in the document]
Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextPredictor.scala]: next pr?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+      batchPerPartition * partitionNum
+    }
+    val otherBroad = dataSet.sparkContext.broadcast(SampleToMiniBatch(
+      batchSize = totalBatch,
+      partitionNum = Some(partitionNum),
+      featurePaddingParam = featurePaddingParam))
+    dataSet.mapPartitions { partition =>
+      val localModel = modelBroad.value()
+      val localTransformer = otherBroad.value.cloneTransformer()
+      val miniBatch = localTransformer(partition)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala]: can we do sth like:
```
miniBatch = localTransformer(partition) ++ Array(null)
miniBatch.flatMap { batch =>
  if (batch != null) {
     ...
  }
  else {
    localModel.release()
    empty
  }
}
```

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+    try {
+      cur = f(preIter.next()).toIterator
+    } catch {
+      case e: Throwable =>
+        shutdownHook()
+        throw e
+    }
+
+  }
+  override def hasNext: Boolean = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala]: What if hasnext is called many times when empty? Can the model be shutdown multiple times?

**Comment by jason-dai** [comment]
Shall we use InferenceModel to build the predictor? @glorysdj @yangw1234

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+    val resultImageFrame = x.toImageFrame() match {
+      case distributedImageFrame: DistributedImageFrame =>
+        Predictor(module, None, batchPerThread)
+          .predictImage(distributedImageFrame, outputLayer = null)
+      case localImageFrame: LocalImageFrame =>
+        val predictor = LocalPredictor(module, None, batchPerCore = batchPerThread)
+        val imageFrame = predictor.predictImage(localImageFrame, outputLayer = null,
+          shareBuffer = false)
+        predictor.shutdown()
```
[the remark this answers is not in the document]
Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/Predictor.scala]: The original bigdl code looks like this. I think maybe they all need to be shutdown.

**Pull request merged**